accepted/2020/serializer/WriteableDomAndDynamic.md

```diff
@@ -0,0 +1,577 @@
+        // statements that need to match typeof(T) to the appropriate JsonElement method.
+        public bool GetBoolean();
+        public byte GetByte();
+        public byte[] GetBytesFromBase64();
+        public System.DateTime GetDateTime();
+        public System.DateTimeOffset GetDateTimeOffset();
+        public decimal GetDecimal();
+        public double GetDouble();
+        public System.Guid GetGuid();
+        public short GetInt16();
+        public int GetInt32();
+        public long GetInt64();
+        public string GetRawText();
+        public sbyte GetSByte();
+        public float GetSingle();
+        public string? GetString();
+        public ushort GetUInt16();
+        public uint GetUInt32();
+        public ulong GetUInt64();
+        public bool TryGetByte(out byte value);
+        public bool TryGetBytesFromBase64(out byte[]? value);
+        public bool TryGetDateTime(out System.DateTime value);
+        public bool TryGetDateTimeOffset(out System.DateTimeOffset value);
+        public bool TryGetDecimal(out decimal value);
+        public bool TryGetDouble(out double value);
+        public bool TryGetGuid(out System.Guid value);
+        public bool TryGetInt16(out short value);
+        public bool TryGetInt32(out int value);
+        public bool TryGetInt64(out long value);
+        public bool TryGetSByte(out sbyte value);
+        public bool TryGetSingle(out float value);
+        public bool TryGetUInt16(out ushort value);
+        public bool TryGetUInt32(out uint value);
+        public bool TryGetUInt64(out ulong value);
+    }
+}
+```
+
+## Serializer interop
+Unlike the 5.0 `JsonNode` prototype and the existing `JsonDocument`, the `JsonNode` types are intended to be used by the serializer directly and do not contain `Parse()` or `Write()` methods themselves because:
+- The existing serializer methods can be used which support UTF8, string, Utf8JsonReader\Writer and Stream. Any future overloads will just work.
+- Serializer-specific functionality including custom converters and quoted numbers just work.
+
+The `JsonSerializerOptions` is passed to `JsonNode` and used to specify any run-time added custom converters, handle other options such as quoted numbers, and to respect the `PropertyNameCaseInsensitive` to determine whether the string-based lookup of property names on `JsonObject` are case-sensitive (for both direct use of `JsonObject` and also indirect use through `dynamic`). If we have `static JsonValue Parse()` methods, those would also need an override to specify the `JsonSerializerOptions`.
+
+Although the serializer can be used to directly (de)serialize `JsonNode` types, once a `JsonNode` instance is created, it is expected the `JsonNode.GetValue<T>()` method is used to change types instead of using the serializer:
+```cs
+    JsonNode node = ... // any value
+    string str;
+
+    // This works, but has no caching:
+    byte[] utf8 = JsonSerializer.SerializeToUtf8Bytes(node, options);
```

Comment:
I would've had a hard time figuring out that to write out the JsonNode I need to use the serializer.

Serializer feels more like a thing for customer-provider types and not something that already represents JSON.

Reply:
The design, like Newtonsoft's, allows any CLR type to be used in "edit mode". Thus the shape of the JSON is not always known.

Reply:
"Serialize()" instance methods were added to JsonNode to make this easier (don't need to pass in `options`) and more discoverable. The existing JsonSerializer.Serialize() methods can still be used however.

However, no "Deserialize()" static methods were added -- the existing JsonSerializer.Deserialize() methods should be used. Note that since a JsonNode can also be a property type or element type of an existing POCO\collection, in many cases no Serialize\Deserialize methods are necessary.

Reply:
> Serializer feels more like a thing for customer-provider types and not something that already represents JSON.

I agree. Why not just `ToJsonString()`?

Reply:
> I agree. Why not just ToJsonString()?

I'm open here. I recently added static Parse() methods and instance-based "To()" methods. However, since they are all "optional" they are in a second-class priority in a way.